File: java/incometax/IncomeTax.java
/** 
 * Pinoy-Sweldo a module for computing Income Tax in the Philippines.
 * :copyright: 2020-Present by Daniel De Castro.
 * :license: MIT, see LICENSE for more details.
*/

package incometax;

import incometax.contribution.*;

public class IncomeTax{

    // calculates the incme tax, netpay, and other contributions

    private SSSContribution sssContribution;
    private PhilHealthContribution philhealthContribution;
    private PagIBIGContribution pagibigContribution;
    private double monthlySalary, totalContributions, taxableIncome, incomeTax, 
        salaryAfterTax, totalDeductions, netPay;

    public IncomeTax(double monthlySalary) throws Exception{
        setSSSContribution(new SSSContribution(monthlySalary));
        setPhilHealthContribution(new PhilHealthContribution(monthlySalary, "employed"));
        setPagibigContribution(new PagIBIGContribution(monthlySalary));
        setMonthlySalary(monthlySalary);
    }

    public SSSContribution getSSSContribution() {
        return sssContribution;
    }

    public void setSSSContribution(SSSContribution sssContribution) {
        this.sssContribution = sssContribution;
    }

    public PhilHealthContribution getPhilHealthContribution() {
        return philhealthContribution;
    }

    public void setPhilHealthContribution(PhilHealthContribution philhealthContribution) {
        this.philhealthContribution = philhealthContribution;
    }

    public PagIBIGContribution getPagibigContribution() {
        return pagibigContribution;
    }

    public void setPagibigContribution(PagIBIGContribution pagibigContribution) {
        this.pagibigContribution = pagibigContribution;
    }

    public double getMonthlySalary() {
        return monthlySalary;
    }

    public void setMonthlySalary(double monthlySalary) {
        this.monthlySalary = monthlySalary;
    }

    public double getTotalContributions() {
        Contribution[] contrib = {getPagibigContribution(), getSSSContribution(), getPhilHealthContribution()};
        double sum = 0;
        for (int i = 0; i < contrib.length; i++){
            sum += contrib[i].getMonthlyContribution();
        }
        totalContributions = sum;
        return totalContributions;
    }

    public void setTotalContributions(double totalContributions) {
        this.totalContributions = totalContributions;
    }

    public double getTaxableIncome() {
        taxableIncome = getMonthlySalary() - getTotalContributions();
        return taxableIncome;
    }

    public void setTaxableIncome(double taxableIncome) {
        this.taxableIncome = taxableIncome;
    }

    public double getIncomeTax() {
        if (getTaxableIncome() <= 20833 )
            incomeTax = 0.0;
        else if (getTaxableIncome() > 20833 && getTaxableIncome() <= 33332 )
            incomeTax = ( getTaxableIncome() - 20833 ) * 0.20;
        else if (getTaxableIncome() == 33333 )
            incomeTax = 2500;
        else if (getTaxableIncome() > 33333 && getTaxableIncome() <= 66666 )
            incomeTax = ( getTaxableIncome() - 33333 ) * 0.25 + 2500;
        else if (getTaxableIncome() == 66667 )
            incomeTax = 10833.33;
        else if (getTaxableIncome() > 66667 && getTaxableIncome() <= 166666 )
            incomeTax = ( getTaxableIncome() - 66667 ) * 0.30 + 10833.33;
        else if (getTaxableIncome() == 166667 )
            incomeTax = 40833.33;
        else if (getTaxableIncome() > 166667 && getTaxableIncome() <= 666666 )
            incomeTax = ( getTaxableIncome() - 166667 ) * 0.32 + 40833.33;
        else if (getTaxableIncome() == 666667 )
            incomeTax = 200833.33;
        else if (getTaxableIncome() > 666667 )
            incomeTax = ( getTaxableIncome() - 666667 ) * 0.35 + 200833.33;
        return incomeTax;
    }

    public void setIncomeTax(double incomeTax) {
        this.incomeTax = incomeTax;
    }

    public double getSalaryAfterTax() {
        salaryAfterTax = getMonthlySalary() - getIncomeTax();
        return salaryAfterTax;
    }

    public void setSalaryAfterTax(double salaryAfterTax) {
        this.salaryAfterTax = salaryAfterTax;
    }

    public double getTotalDeductions() {
        totalDeductions = getIncomeTax() + getTotalContributions();
        return totalDeductions;
    }

    public void setTotalDeductions(double totalDeductions) {
        this.totalDeductions = totalDeductions;
    }

    public double getNetPay() {
        netPay = getMonthlySalary() - getTotalDeductions();
        return netPay;
    }

    public void setNetPay(double netPay) {
        this.netPay = netPay;
    }

    public String toRepresentation(){
        return "Monthly salary = "+getMonthlySalary()+"\n"
            + getSSSContribution().getModel()+" = "+getSSSContribution().getMonthlyContribution()+"\n"
            + getPhilHealthContribution().getModel()+" = "+getPhilHealthContribution().getMonthlyContribution()+"\n"
            + getPagibigContribution().getModel()+" = "+getPagibigContribution().getMonthlyContribution()+"\n"
            + "Total contributions = "+getTotalContributions()+"\n"
            + "Income tax = "+getIncomeTax()+"\n"
            + "Salary after tax = "+getSalaryAfterTax()+"\n"
            + "Total deductions = "+getTotalDeductions()+"\n"
            + "Net pay = "+getNetPay()+"\n";
    }
    
}

File: python/incometax/income_tax.py
from pandas import DataFrame
from .contributions import PhilHealthContribution, SSSContribution, PagIBIGContribution

class IncomeTax:
    
    def __init__(self, monthly_salary):
        self.sss_contribution = SSSContribution(monthly_salary)
        self.philhealth_contribution = PhilHealthContribution(monthly_salary)
        self.pagibig_contribution = PagIBIGContribution(monthly_salary)
        self.monthly_salary = monthly_salary
    
    @property
    def total_contributions(self):
        return sum([c.monthly_contribution for c in (
            self.philhealth_contribution,
            self.sss_contribution,
            self.pagibig_contribution
        )])

    @property
    def taxable_income(self):
        return self.monthly_salary - self.total_contributions

    def set_philhealth_membership(self, membership = 'employed'):
        self.philhealth_contribution = PhilHealthContribution(self.monthly_salary, membership)

    @property
    def income_tax(self):
        taxable_income = self.taxable_income
        if taxable_income <= 20833:
            return 0.0
        elif taxable_income > 20833 and taxable_income <= 33332:
            return ( taxable_income - 20833 ) * 0.20
        elif taxable_income == 33333:
            return 2500
        elif taxable_income > 33333 and taxable_income <= 66666:
            return ( taxable_income - 33333 ) * 0.25 + 2500
        elif taxable_income == 66667:
            return 10833.33
        elif taxable_income > 66667 and taxable_income <= 166666:
            return ( taxable_income - 66667 ) * 0.30 + 10833.33
        elif taxable_income == 166667:
            return 40833.33
        elif taxable_income > 166667 and taxable_income <= 666666:
            return ( taxable_income - 166667 ) * 0.32 + 40833.33
        elif taxable_income == 666667:
            return 200833.33
        elif taxable_income > 666667:
            return ( taxable_income - 666667 ) * 0.35 + 200833.33
    
    @property
    def salary_after_tax(self):
        return self.monthly_salary - self.income_tax
    
    @property
    def total_deductions(self):
        return self.income_tax + self.total_contributions
    
    @property
    def net_pay(self):
        return self.monthly_salary - self.total_deductions

    def __repr__(self):
        data = [
            ['Monthly salary', self.monthly_salary],
            ['SSS', self.sss_contribution.monthly_contribution ],
            ['PhilHealth', self.philhealth_contribution.monthly_contribution],
            ['PAGIBIG', self.pagibig_contribution.monthly_contribution],
            ['Total Contributions', self.total_contributions],
            ['Income Tax', self.income_tax],
            ['Salary after tax', self.salary_after_tax],
            ['Total deductions', self.total_deductions],
            ['Net pay', self.net_pay]
        ]
        return str(DataFrame(data, columns=['---','Php']))
